airbyte-workers/src/main/java/io/airbyte/workers/temporal/sync/LauncherWorker.java

```diff
@@ -117,12 +119,30 @@ public OUTPUT run(INPUT input, Path jobRoot) throws WorkerException {
             containerOrchestratorConfig.containerOrchestratorImage(),
             containerOrchestratorConfig.googleApplicationCredentials());
 
+        cancellationCallback.set(() -> {
+          // When cancelled, try to set to true.
+          // Only proceed if value was previously false, so we only have one cancellation going. at a time
+          if (!isCanceled.getAndSet(true)) {
+            log.info("Trying to cancel async pod process.");
+            process.destroy();
+          }
+        });
+
+        // only kill running pods and create process for the first run for an attempt
         if (process.getDocStoreStatus().equals(AsyncKubePodStatus.NOT_STARTED)) {
-          process.create(
-              allLabels,
-              resourceRequirements,
-              fileMap,
-              portMap);
+          log.info("Creating " + podName + " for attempt number: " + jobRunConfig.getAttemptId());
+          killRunningPodsForConnection(podName);
```

Comment:
my understanding from our conversation about state machine yesterday was that we 

* determine what pod we want to run
* kill any running pod (for the connection) that doesn't match it
* then if no pod is running that matches out running pod, start a pod

my understanding from this code change is the killing is now only happening in this conditional statement so the case where there is something running that doesn't match what we are looking for, then it won't get killed? is that what we want? am i misunderstanding?

Reply:
It's more:
* Determine what pod we want to run
* If we haven't successfully initialized this pod in the past, kill any pod for that connection and create. If we can't create fail the attempt.
* Then attach to the pod (whether or not it was initialized this run)

Since we always run the deletion when creating, it eliminates the risk of our filtering down to a specific podName having an error in some way; it just axes everything for that connection id.

Do you think that's sufficiently close to the state machine diagram or should we change it on one side?

Reply:
yup. i'm convinced.